Removed the built-in overlay from ReorderableListView.

darrenaustin opened the pull request:
The current implementation of `ReorderableListView` creates its own `Overlay` widget to handle the dragging of its items. However, this is causing layout problems for apps that are trying to use `shrinkWrap`, as the overlay is not constrained.
This PR removes this and has the ReorderableListView depend on the surrounding overlay provided by MaterialApp. If one isn't found it will throw an error.

Fixes: #76150

I added a new test to verify an error if no surrounding overlay is found. Also removed two golden tests that would need to be updated as the dragged item is no longer painting in the list itself (they were of no real testing value either).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

darrenaustin committed:
Removed the built-in overlay from the ReorderableListView so that it now depends on the surrounding overlay.

darrenaustin commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -289,115 +289,7 @@ class ReorderableListView extends StatefulWidget {
   _ReorderableListViewState createState() => _ReorderableListViewState();
 }
 
-// This top-level state manages an Overlay that contains the list and
```
This looks more drastic than it is. The previous implementation of _RerorderableListViewState wrapped  `_ReorderableListContent` with an overlay that it managed. This is no longer needed, so this just removes the `_ReorderableListContent` widget and replaces the _ReorderableListViewState with the contents of the old `_ReorderableListContentState` which is where the real list state implementation was.

goderbauer commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -212,47 +212,6 @@ void main() {
         expect(getListHeight(), kDraggingListHeight);
       });
 
-      testWidgets('Vertical drop area golden', (WidgetTester tester) async {
```
What did these tests verify? Why do we no longer need to guarantee that this use case works?

darrenaustin replied:
It is not entirely clear from the code, but from the title 'Vertical drop area golden', I presume it was trying to test that there was a drop area when an item was dragged. However, looking at this code it creates a three item list, starts dragging the middle item (blue) and then takes a screen shot. However, as the item has not yet moved, it is sitting over the drop area, so the golden image is just a list with three items in it. Not sure what it was trying to verify here.  Removing the overlay would mean the item is no longer in the list view's repaint boundary, so the golden would have to be updated, but again, I am not sure what value it provides.

goderbauer replied:
> However, as the item has not yet moved, it is sitting over the drop area, so the golden image is just a list with three items in it.

Is that maybe what the test was testing? If you pick up an item it starts out to render at the correct location (i.e. at its original location in the list).

darrenaustin replied:
Yeah, as I was explaining this above, that thought occurred to me as well. I will update the test to show the missing item during the drag and update the golden for it. Thx for the input.

darrenaustin replied:
I just pushed a commit that re-added these tests, but updated them for capturing a drag in progress.

goderbauer reviewed (commented):
I thought there was a reason why we added the overlay (e.g. to make sure that you can't drag an item outside of the list). Why do we not need that anymore?

darrenaustin commented:
> I thought there was a reason why we added the overlay (e.g. to make sure that you can't drag an item outside of the list). Why do we not need that anymore?

As far as I remember I had removed it, but we put it back near the end of development as it was causing some tests to fail or some other reason.

I don't think having a built-in overlay prevents the item from being dragged outside the list, as it doesn't appear to be clipping it (and I am not sure we would want it to from a UI perspective). The list already handles the case of dragging outside the list and just puts the gap at the edge of the list and the item will animate back to that gap if dropped outside the list.

In addition, it is inconsistent with the other two reorderable list widgets as they both rely on the surrounding overlay widget.

Therefore I don't see much value in keeping the build-in overlay here as it provides little value and is breaking the shrink-wrap case as it doesn't shrink wrap around the list. 

Or am I missing something here?

goderbauer commented:
> I don't think having a built-in overlay prevents the item from being dragged outside the list, as it doesn't appear to be clipping it (and I am not sure we would want it to from a UI perspective).

I thought there was a comment in the code explaining why there is an internal Overlay, but looks like it got lost during the refactor. Prior to the refactor, this is what it said:

https://github.com/flutter/flutter/blob/ff05ca2503437ca35ea07551aeda5a6b1d5d35bb/packages/flutter/lib/src/material/reorderable_list.dart#L175

And looking at the Overlay I would expect it to be clipping (it's the default value for https://master-api.flutter.dev/flutter/widgets/Overlay/clipBehavior.html and we don't seem to change it...). Is that not working?

Whether not drawing outside the list itself is a bug or a feature, I actually don't know... The more I think about it the more I am leaning towards it should draw outside of the list. So, if that was the only reason, maybe we should remove it...

darrenaustin commented:
> And looking at the Overlay I would expect it to be clipping (it's the default value for https://master-api.flutter.dev/flutter/widgets/Overlay/clipBehavior.html and we don't seem to change it...). Is that not working?
> 

I just tried it and it seemed to paint the dragged item over the app bar and such just fine. I will take another look to see if I was just holding wrong or something.

> Whether not drawing outside the list itself is a bug or a feature, I actually don't know... The more I think about it the more I am leaning towards it should draw outside of the list. So, if that was the only reason, maybe we should remove it...

That's pretty much the conclusion that led me to this PR. If you do come up with a case for the built-in overlay, just let me know.

darrenaustin committed:
Put the removed golden tests back, but cleaned them up and updated them to show a drag in progress.

darrenaustin commented:
> I just tried it and it seemed to paint the dragged item over the app bar and such just fine. I will take another look to see if I was just holding wrong or something.

Ok, I checked again, and I was holding it wrong. It does clip, but as mentioned before I don't think that is the behavior we want here. It is inconsistent with the other reorderable list widgets and if someone really wants the dragged item to clip, they can wrap it with their own overlay.

skia-gold commented:
Gold has detected about 6 untriaged digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/79024

goderbauer commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -212,7 +212,8 @@ void main() {
 
-      testWidgets('Vertical drop area golden', (WidgetTester tester) async {
+      testWidgets('Vertical drag in progress golden image', (WidgetTester tester) async {
+        debugDisableShadows = false;
```
I believe this has to be reset to its original value at the end of the test?

goderbauer commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -1342,6 +1376,38 @@ void main() {
     expect(exception, isFlutterError);
     expect(exception.toString(), contains('Every item of ReorderableListView must have a key.'));
   });
+
```
Should we also include a test where the Overlay does not tightly wrap the list (i.e. the list has to share the screen space with another sibling) to ensure that our math is correct to the origin of the overlay and not the origin of the list?

goderbauer reviewed (approved):
LGTM

goderbauer commented:
> as mentioned before I don't think that is the behavior we want here. It is inconsistent with the other reorderable list widgets and if someone really wants the dragged item to clip, they can wrap it with their own overlay.

Sounds good to me.

darrenaustin committed:
Reset the debugDisableShadows at the end of the golden tests.

darrenaustin committed:
Added padding to test when the origin of the overlay doesn't match the origin of the list.
